syntax = "proto3";

package gradeservice;

service GradeService {
  // Get grades for a specific student
  rpc GetStudentGrades(GetGradesRequest) returns (GradesResponse);
  
  // Upload grades for a course (internal gRPC only)
  rpc UploadGrades(UploadGradesRequest) returns (UploadGradesResponse);
  
  // Stream grades for a student (for large datasets)
  rpc StreamStudentGrades(GetGradesByStudentRequest) returns (stream GradeRecord);
}

message GetGradesRequest {
  int32 student_id = 1;
}

message GradeRecord {
  int32 id = 1;
  int32 student_id = 2;
  int32 course_id = 3;
  string grade_value = 4;
  string course_code = 5;
  string course_name = 6;
  int64 uploaded_at = 7;
  int32 uploaded_by = 8;
}

message GradesResponse {
  repeated GradeRecord grades = 1;
  int32 count = 2;
}

message UploadGradesRequest {
  int32 course_id = 1;
  int32 uploaded_by = 2;
  repeated GradeEntry entries = 3;
}

message GradeEntry {
  int32 student_id = 1;
  string grade_value = 2;
}

message UploadGradesResponse {
  int32 created_count = 1;
  bool success = 2;
  string message = 3;
}

message GetGradesByStudentRequest {
  int32 student_id = 1;
}
